pymtl3/dsl/Component.py

```diff
@@ -55,10 +61,543 @@ def _construct( s ):
+              obj._dsl.param_tree.merge( node )
+
+            elif node.compiled_re is not None:
+              if node.compiled_re.match( u_name ):
+                # Lazily create the param tree
+                if obj._dsl.param_tree is None:
+                  obj._dsl.param_tree = ParamTreeNode()
+                obj._dsl.param_tree.merge( node )
+
+      obj._dsl.full_name = ( parent._dsl.full_name + "." + u_name )
+
+      # store the name/indices
+      obj._dsl._my_name     = name
+      obj._dsl._my_indices  = indices
+
+      NamedObject.__setattr__ = NamedObject.__setattr_for_elaborate__
+      obj._construct()
+      del NamedObject.__setattr__
+
+    top = s._dsl.elaborate_top
+
+    added_components = obj._collect_all( [ lambda x: isinstance( x, Component ) ] )[0]
+
+    # First elaborate all functions to spawn more named objects
+    for c in added_components:
+      c._dsl.elaborate_top = top
+      c._elaborate_read_write_func()
+
+    added_signals, added_method_ports = \
+      obj._collect_all( [ lambda x: isinstance( x, Signal ), \
+                          lambda x: isinstance( x, MethodPort ) ] )
+
+    top._dsl.all_components    |= added_components
+    top._dsl.all_signals       |= added_signals
+    top._dsl.all_method_ports  |= added_method_ports
+
+    top._dsl.all_named_objects |= added_components
+    top._dsl.all_named_objects |= added_signals
+    top._dsl.all_named_objects |= added_method_ports
+
+    for c in added_components:
+      top._collect_vars( c )
+
+    for c in added_signals | added_method_ports:
+      c._dsl.elaborate_top = top
+
+    # Lazy -- to avoid resolve_connection call which takes non-trivial
+    # time upon adding any connect, I just mark it here. Please make sure
+    # to call s.get_all_value_nets() to flush all pending connections
+    # whenever you want to get the nets
+    connection_pairs = []
+    for (x, y) in provided_connections:
+      connection_pairs.append( x )
+      connection_pairs.append( eval(y) )
+
+      if not top._dsl._has_pending_value_connections and isinstance( x, Signal ):
+        top._dsl._has_pending_value_connections = True
+      if not top._dsl._has_pending_method_connections and isinstance( x, MethodPort ):
+        top._dsl._has_pending_method_connections = True
+
+    top.add_connections( *connection_pairs )
+
+  def _delete_component( top, obj ):
+
+    # This nested delete function is to create an extra layer to properly
+    # call garbage collector. If we can make sure it is collected
+    # automatically and fast enough, we might remove this extra layer
+    #
+    # EDIT: After experimented w/ and w/o gc.collect(), it seems like it
+    # is eventually collected, but sometimes the intermediate memory
+    # footprint might reach up to gigabytes, so let's keep the
+    # gc.collect() for now
+
+    def _delete_component_internal( top, foo ):
+      is_placeholder = isinstance( obj, Placeholder )
```

Comment:
Seems like you are not assuming `Placeholder` here.  What's this flag for?

Reply:
Hmmm good catch. There is one line below that checks isinstance(obj,Placeholder) again. The purpose was to check if it is placeholder. If it is, I don't need to call those uncollect thingy.